lib/modules/asset/AssetService.ts

```diff
@@ -304,213 +303,105 @@ export class AssetService {
-            assetsList
-          );
-        });
-
-        // Delete the assets in the old tenant
-        await this.sdk.document.mDelete(
-          engineId,
-          InternalCollection.ASSETS,
-          assetsList
-        );
 
-        recovery.addRecovery(async () => {
-          await this.sdk.document.mCreate(
-            engineId,
-            InternalCollection.ASSETS,
-            assetsToMigrate
-          );
-        });
-
-        // Link the devices to the new assets
-        const promises: Promise<void>[] = [];
-
-        for (const asset of assetLinkedDevices) {
-          const assetId = asset.assetId;
-          for (const device of asset.linkedDevices) {
-            const deviceId = device._id;
-            const measureNames = device.measureNames;
-            promises.push(
-              ask<AskDeviceLinkAsset>("ask:device-manager:device:link-asset", {
-                assetId,
-                deviceId,
+            // ... and link this device to the asset in the new tenant
+            await ask<AskDeviceLinkAsset>(
+              "ask:device-manager:device:link-asset",
+              {
+                assetId: asset._id,
+                deviceId: device.id,
                 engineId: newEngineId,
-                measureNames: measureNames,
+                measureNames: device.measureNames,
                 user,
-              })
+              }
             );
           }
-        }
 
-        await Promise.all(promises);
-
-        recovery.addRecovery(async () => {
-          const promiseRecoveries: Promise<void>[] = [];
-
-          for (const asset of assetLinkedDevices) {
-            for (const device of asset.linkedDevices) {
-              const deviceId = device._id;
-              promiseRecoveries.push(
-                ask<AskDeviceUnlinkAsset>(
-                  "ask:device-manager:device:unlink-asset",
-                  {
-                    deviceId,
-                    user,
-                  }
-                )
-              );
-            }
-          }
-
-          await Promise.all(promiseRecoveries);
-        });
+          // Finally here, we can delete the asset in the source engine !
+          await this.sdk.document.delete(
+            engineId,
+            InternalCollection.ASSETS,
+            asset._id
+          );
 
-        // clear the groups
-        await this.sdk.document.mUpdate<AssetContent>(
-          newEngineId,
-          InternalCollection.ASSETS,
-          assetsList.map((assetId) => ({
-            _id: assetId,
-            body: {
-              groups: [],
-            },
-          }))
-        );
+          migrated++;
+        }
       } catch (error) {
-        await recovery.rollback();
         throw new BadRequestError(
           `An error occured while migrating assets: ${error}`
         );
       }
+
+      if (!migrated) {
+        throw new BadRequestError(
+          `Error occured while migrating all the assets !`
```

Comment:
Instead of throwing this error, you can print all the assets that havent been migrated

Reply:
And maybe add the one the succeeded so we can print in the interface to the user the complete result of the process

Reply:
You can write a tests that fail to test that you print the right data

Reply:
In that precise cas, I want to throw an error when ALL assets have failed to migrate.
An improvement should be to return errors [] / successes [] to know which ones have been successfully migrated, and the ones that failed.
For the test, yes ... but it will spend lots of hours to me ;-)

Reply:
If you don't want to take the time now to have such arrays to print errors and success migration, you should add an issue in this repository that talk about this improvements that should be done.

Also we should not discuss about how much time it will take to add a test, because if at some point we have to change this function behaviour and we made a regression about this part of the code. It will be much harder to debug in 6 months because we will have forgotten what we written here. leading to much more hours in debugging that the time spent to write a single test

Reply:
I'll add an issue for this, in order to implement it later, in a next iteration